### PR TITLE
51423: wp-admin/setup-config.php fixes fwrite handle mismatch

### PR DESCRIPTION
**Pull request by hellofromtonya**
Trac ticket: https://core.trac.wordpress.org/ticket/51423

- Fixes a `fwrite` handle type mismatch which happens if [`fopen`](https://www.php.net/manual/en/function.fopen.php) fails, i.e. returns `false`
- Adds error messaging for it

Note:
The new conditional uses the absence of `false` instead of checking for a resource. Why? PHP has plans to change `fopen` to return `object` instead of current `resource` in a future version. This not falsey check though opposite ensures the code is future-proof when the new version is released. 

An inline comment (thanks @peterwilsoncc) is added to explain and provides a learning opportunity for those who read the code.

@props xknown, hellofromTonya, jrf, peterwilsoncc

## TODO

- [x] Review new error messaging with the Polyglots team
- [x] Review `chmod` failed error messaging

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [inline src/wp-admin/setup-config.php:441]:
```suggestion
		if ( false !== $handle ) {
```

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [inline src/wp-admin/setup-config.php:446]: Needs to catch if a failure happens to alert that the config is not going to be written, i.e. similar to:

```
			<?php
			/* translators: %s: wp-config.php */
			printf( __( 'Unable to write to %s file.' ), '<code>wp-config.php</code>' );
			?>
```

**Comment by hellofromtonya** [comment]
Force push to reset the file back to original state (except for the indentation). Why? To first show the problem exists in PHP 8, but not in the other PHP versions. Doing this step shows the problem in action for crafting the right fix without causing a breaking change.

**Comment by hellofromtonya** [comment]
## Manual Testing BEFORE the Fix

Environment:
- Using Local as the localhost app
- Nginx
- MySQL 8.0.16
- OS: Mac Big Sur
- Browser: Firefox

To make it testable, needed to create an environment where `fopen` fails to create the file: 

- Changed the permissions of the existing `wp-config.php` to `000`
```bash
----------  1 ****  staff  3181 Mar 15 13:25 wp-config.php
```
- Skipped `file_exists` check by commenting the check on lines 58 to 69
- Flushed the realpath cache for the file
```php
clearstatcache( ABSPATH . 'wp-config.php' );
```

### Test 1: with PHP 7.3.5

Results: Works, but PHP Warning

From the error.log
```
PHP Warning:  fwrite() expects parameter 1 to be resource, bool given in .../setupconfig/app/public/wp-admin/setup-config.php on line 443
```

PR722-Before-PHP7

### Test 2: with PHP 8.0

Results: Does not work due to `TypeError` fatal error

From the error.log
```
PHP Fatal error:  Uncaught TypeError: fwrite(): Argument #1 ($stream) must be of type resource, bool given in .../setupconfig/app/public/wp-admin/setup-config.php:443
Stack trace:
#0 .../setupconfig/app/public/wp-admin/setup-config.php(443): fwrite(false, '<?php\r\n')
#1 {main}
  thrown in .../setupconfig/app/public/wp-admin/setup-config.php on line 443
  ```

PR722-before-php8

**Comment by hellofromtonya** [comment]
## Manual Testing AFTER the Fix

Same environment and test conditions.

### Test 1: with PHP 7.3.5

Results: Works with no breaking change.

- `wp-config.php` file permissions is changed
- Error message displays as expected
- Error Log:
    - No `fwrite` PHP Warning
    - PHP Warning for `fopen` as expected: `PHP Warning:  fopen(/.../setupconfig/app/public/wp-config.php): Failed to open stream: Permission denied`

pr722-after-php7 3 8

### Test 2: with PHP 8.0

Results: Works.

- `wp-config.php` file permissions is changed
- Error message displays as expected
- Error log:
   - No PHP `TypeError` Fatal Error
   - PHP Warning for `fopen` as expected: `PHP Warning:  fopen(/.../setupconfig/app/public/wp-config.php): Failed to open stream: Permission denied`

pr722-after-php8

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [inline src/wp-admin/setup-config.php:453]:
```suggestion
							__( 'Unable to write to %1$s file due  to too restrictive file permissions: %2$s. Please update the chmod file permissions to allow for write access.' ),
```

Ask in the polyglot channel to validate wording.

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [inline src/wp-admin/setup-config.php:463]:
```suggestion
			if  ( ! chmod( $path_to_wp_config, 0666 ) ) {
				// Add error message.
			}
```

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [reply src/wp-admin/setup-config.php]: Code updated in commit https://github.com/WordPress/wordpress-develop/pull/722/commits/c6d272a45e45d97830b04b218b423b0d0a4a6e28

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [reply src/wp-admin/setup-config.php]: Code updated in commit https://github.com/WordPress/wordpress-develop/pull/722/commits/c6d272a45e45d97830b04b218b423b0d0a4a6e28

**Comment by hellofromtonya** [comment]
[Review in slack by polyglots](https://wordpress.slack.com/archives/C02RP50LK/p1617294051346100) ✅

**Review by jrfnl — APPROVED** [review]
LGTM. Approved. Let's wait for a second opinion from the security team before submitting it for commit as we're touching the messaging to the end-user with potentially security sensitive information.

File: src/wp-admin/setup-config.php
Comment (peterwilsoncc) [inline src/wp-admin/setup-config.php:455]: I'm in two minds about including the current permission, happy to say the file is not writable as the visitor can install at this point. The thought behind the question is in case the site's owners messed up and made the file neither readable or writable by the web server user, eg `0600`, in which case it reveals something about the experience level of the person configuring the server.

In most cases, WP links to docs so that might help too:

https://github.com/WordPress/wordpress-develop/blob/2d7b7dd833070386525ac04af45d5e64d10b8937/src/wp-admin/includes/file.php#L321-L327

Maybe:

```php
printf(
	/* translators: 1 wp-config.php, 2: Documentation URL. */
	__( 'You need to make the file %1$s writable before you can save your changes. See <a href="%2$s">Changing File Permissions</a> for more information.' ),
	__( 'https://wordpress.org/support/article/changing-file-permissions/' )
);
```

As I say, I am in two minds so consider this a nit.

**Review by peterwilsoncc — COMMENTED** [review]
Reviewed this with whitespace only changes suppressed.

These changes look good to me, a couple of comments/questions inline but nothing I consider a blocker.

File: src/wp-admin/setup-config.php
Comment (jrfnl) [reply src/wp-admin/setup-config.php:442]: The reason for not using `is_resource()` is that it is already known that the PHP project plans to change the return type of the `fopen()` function from `resource|false` to `object|false`.

Comparing against `!== false` future-proofs this code snippet for that change.

See: https://github.com/php/php-tasks/issues/6

Comment (hellofromtonya) [reply src/wp-admin/setup-config.php:446]: FYI: This comment will need to updated once the PHP version is released that has this change in it.

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [reply src/wp-admin/setup-config.php:463]: @peterwilsoncc This string `Unable to write to %s file.` is a copy from another one earlier in the same file:

https://github.com/WordPress/wordpress-develop/blob/c072116d6aa3f182e6d91dea19ce76c59f7d16d5/src/wp-admin/setup-config.php#L399-L402

Are you suggesting to change both strings?

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [reply src/wp-admin/setup-config.php]: I agree with you @peterwilsoncc. That string reads better and provides better guidance for users. Updated in commit https://github.com/WordPress/wordpress-develop/pull/722/commits/89d6689fd708fbe1e779a3b6dfcd6e1057ac6d8f

File: src/wp-admin/setup-config.php
Comment (hellofromtonya) [reply src/wp-admin/setup-config.php:445]: After discussing with @jrfnl, removed the `@see` link from the inline comment in commit https://github.com/WordPress/wordpress-develop/pull/722/commits/34d3376648d6dd43d6c9a0901ef4e6d350575cdc

File: src/wp-admin/setup-config.php
Comment (peterwilsoncc) [inline src/wp-admin/setup-config.php:444]: Sorry, multi-line comment style nit.

```suggestion
			/*
			 * Why check for the absence of false instead of checking for resource with is_resource()?
			 * To future-proof the check for when fopen returns object instead of resource, i.e. a known
			 * change coming in PHP.
			 */
```

File: src/wp-admin/setup-config.php
Comment (peterwilsoncc) [inline src/wp-admin/setup-config.php:456]:
```suggestion
						'<code>wp-config.php</code>',
						__( 'https://wordpress.org/support/article/changing-file-permissions/' )
```

**Review by peterwilsoncc — COMMENTED** [review]
Sorry, a couple more things.

Re: the string `'Unable to write to %s file.'`, no need to change if it's from another file. I must have done a silly when searching, probably had regex mode enabled or some such.

**Review by peterwilsoncc — APPROVED** [review]
Lovely! I'll commit these changes today.

Reviewed with whitespace changes suppressed.

**Comment by peterwilsoncc** [comment]
Committed in https://core.trac.wordpress.org/changeset/50775

**Pull request closed without merging**
